eval/topdown.go

```diff
@@ -159,21 +194,69 @@ var builtinFunctions = map[opalog.Var]builtinFunction{
 // dereferenceVar is used to lookup the variable binding and convert the value to
 // a native Go type.
 func dereferenceVar(v opalog.Var, ctx *TopDownContext) (interface{}, error) {
-	binding, ok := ctx.Bindings[v]
-	if !ok {
+	binding := ctx.Bindings.Get(v)
+	if binding == nil {
 		return nil, fmt.Errorf("unbound variable: %v", v)
 	}
-
 	return valueToInterface(binding, ctx)
 }
 
+func evalContext(ctx *TopDownContext, iter TopDownIterator) error {
+
+	if ctx.Index >= len(ctx.Query) {
+		ctx.traceFinish()
+		return iter(ctx)
+	}
+
+	ctx.traceEval()
+
+	if ctx.Current().Negated {
+		return evalContextNegated(ctx, iter)
+	}
+
+	return evalTerms(ctx, func(ctx *TopDownContext) error {
```

Comment:
Thanks for the rundown.  I hear that go-routines are super expensive (incurring a context switch on each one).  So let's stay away from that one.  Callbacks seem fine.